### PR TITLE
Test Refactor and more "kit" command tests

### PR DESCRIPTION
**Pull request by meenakb**
Kit command tests

### CODE REVIEW

File: src/taco-cli/test/kit.ts [hunk 1]
```diff
@@ -97,11 +108,30 @@ describe("Kit", function (): void {
...
+        .then(function (): Q.Promise<any> {
+                throw new Error("Scenario succeeded when it should have failed");
+        }, function (err: tacoUtils.TacoError): Q.Promise<any> {
+            err.errorCode.should.equal(expectedErrorCode);
+
+            // Also make sure that the project's taco.json
+            // file exists and has the expected values
+
+            fs.existsSync(tacoJsonPath).should.be.true;
+            var tacoJson: IKeyValuePair<string> = require(tacoJsonPath);
```
Comment (MSLaguana) [inline src/taco-cli/test/kit.ts]: Will tacoJsonPath ever be the same between two different tests in the same run? If so, I'd recommend against using `require` since that does caching based on file paths. To make sure you get the file as it exists on disk it may be better to explicitly `JSON.parse(fs.readFileSync(tacoJsonPath))`

**Comment by joshuaskelton** [comment]
Looks good to me.

File: src/taco-cli/test/kit.ts [hunk 2]
```diff
@@ -97,11 +108,30 @@ describe("Kit", function (): void {
 
             var tacoJson: IKeyValuePair<string> = require(tacoJsonPath);
 
-            tacoJsonKeyValues.should.be.eql(tacoJson);
+            tacoJsonKeyValues.should.be.eql(tacoJson)
```
Comment (digeff) [inline src/taco-cli/test/kit.ts]: Why did you remove the ; ?

File: src/taco-cli/test/kit.ts [hunk 3]
```diff
@@ -198,8 +225,13 @@ describe("Kit", function (): void {
             rimraf(kitProjectpath, function (err: Error): void { done(); }); // ignore errors
         });
 
+        it("'taco kit select --cordova {Invalid-CLI-VERSION}' should execute with expected errors", function (done: MochaDone): void {
+            runKitCommandFailureCaseAndVerifyTacoJson<TacoErrorCodes>(["select", "--cordova", "InvalidCordovaCliVersion"], tacoJsonPath, expectedKitTacoJsonKeyValues, TacoErrorCodes.ErrorInvalidVersion)
+            .done(() => done(), done)
```
Comment (digeff) [inline src/taco-cli/test/kit.ts]: Isn't this equivalent to .done(done, done) ?

Comment (MSLaguana) [inline src/taco-cli/test/kit.ts]: No; `.done(done, done)` would pass through whatever arguments in the first case, so for example `[...].then(() => foo).done(done, done)` would invoke the first done with `foo` (which indicates an error condition). By doing `.done(() => done(), done)` instead we are explicitly invoking done with no arguments in the success case.

However last week I was reading the mocha documentation and noticed that we can avoid all of this by simply returning the promise, and mocha will succeed on a resolved promise and fail on a rejected promise.

Comment (digeff) [inline src/taco-cli/test/kit.ts]: Nice! I like the returning promise way!

File: src/taco-cli/test/kit.ts (hunk 1 above)
Comment (meenakb) [inline src/taco-cli/test/kit.ts]: That's a good point. Although the project paths are different, fixed it.

File: src/taco-cli/cli/kit.ts [hunk 4]
```diff
@@ -95,8 +104,9 @@ class Kit extends commands.TacoCommandBase {
         return Kit.promptUser(resources.getString("CommandKitSelectProjectUpdatePrompt"))
         .then(function (answer: string): Q.Promise<any> {
             if (answer && answer.length > 0 ) {
-                answer = answer.toLowerCase();
-                if (resources.getString("PromptResponseYes").split("\n").indexOf(answer) !== -1) {
+                // Try the lower-case version of the answer as well (This is needed so we can have better test coverage for "yes" and "no" response code paths)
```
Comment (MSLaguana) [inline src/taco-cli/cli/kit.ts]: I don't follow this comment; why do we need to check both original-case and lower-case? why not just always convert to lower case as we did before?

File: src/taco-cli/test/kit.ts [hunk 5]
```diff
@@ -64,6 +73,22 @@ describe("Kit", function (): void {
...
+
+    var expectedCliTacoJsonKeyValues2: IKeyValuePair<string> = {
+        "cordova-cli": "5.1.1"
+    };
+
+    var expectedKitTacoJsonKeyValues: IKeyValuePair<string> = {
+        kit: "5.1.1-Kit", "cordova-cli": "5.1.1"
+    };
+
+    var expectedKitPlatformVersion: IComponentVersionMap = {
```
Comment (MSLaguana) [inline src/taco-cli/test/kit.ts]: Why does this have a different type to the others? looks like it is actually equivalent, is that right?

File: src/taco-cli/cli/kit.ts (hunk 4 above)
Comment (meenakb) [inline src/taco-cli/cli/kit.ts]: In the test mode, we do not resolve the actual strings from resources.json (resouces.getString("PromptResponseYes") will just return "PromptResponseYes"  which means that in order to test the positive code paths, we will have to pass "PromptResponseYes" (for this check to succeed). The goal was to get as much test coverage as possible.

File: src/taco-cli/test/kit.ts [hunk 6]
```diff
@@ -89,19 +114,96 @@ describe("Kit", function (): void {
...
+    function addPlatformToProject(platfromName: string, projectPath: string): Q.Promise<any> {
+        process.chdir(projectPath);
+        return platformRun(["add", platfromName]);
+    }
+
+    function addTestPluginsToProject(projectPath: string): Q.Promise<any> {
+        process.chdir(projectPath);
+        return pluginRun(["add", "cordova-plugin-camera"])
+        .then(function(): Q.Promise<any> {
+            return sleep(100);
```
Comment (MSLaguana) [inline src/taco-cli/test/kit.ts]: Why do we need a sleep here?

File: src/taco-cli/cli/kit.ts (hunk 4 above)
Comment (MSLaguana) [inline src/taco-cli/cli/kit.ts]: Ah, I see. In that case I'd be inclined to call toLowerCase on the getString() response as well, to make sure we are comparing both sides as lower case. I see that as less of a "making allowances for test scenarios" and more "protecting us against small mistakes in (possibly localized) strings", which also happens to mean we can work in a test scenario

File: src/taco-cli/test/kit.ts [hunk 7]
```diff
@@ -89,19 +114,96 @@ describe("Kit", function (): void {
...
+                        break;
+                    default:
+                        errorHandler(new Error("Unexpected query!"));
+                }
+            },
+           close: onClose
+        };
+    }
+
+    function sleep(milliseconds: number): Q.Promise<any> {
```
Comment (MSLaguana) [inline src/taco-cli/test/kit.ts]: You can use Q.delay(milliseconds) for this

File: src/taco-cli/test/kit.ts (hunk 5 above)
Comment (meenakb) [inline src/taco-cli/test/kit.ts]: Yes :) This can as well be of the same type.

File: src/taco-cli/test/kit.ts (hunk 7 above)
Comment (meenakb) [inline src/taco-cli/test/kit.ts]: Q.delay(milliseconds) seems like a better alternative!

File: src/taco-cli/test/kit.ts [hunk 8]
```diff
@@ -89,19 +114,96 @@ describe("Kit", function (): void {
...
-            var tacoJson: IKeyValuePair<string> = require(tacoJsonPath);
-
+            var tacoJson: IKeyValuePair<string> = JSON.parse(fs.readFileSync(tacoJsonPath, { encoding: "utf-8" }));
             tacoJsonKeyValues.should.be.eql(tacoJson);
             return telemetryParameters;
         });
     }
 
+    function runKitCommandFailureCaseAndVerifyTacoJson<T>(args: string[],
+        tacoJsonPath: string, tacoJsonKeyValues: IKeyValuePair<string>, expectedErrorCode: T): Q.Promise<TacoUtility.ICommandTelemetryProperties> {
```
Comment (MSLaguana) [inline src/taco-cli/test/kit.ts]: This function doesn't really return a Q.Promise<TacoUtility.ICommandTelemetryProperties> since it is only the error case. Perhaps Q.Promise<void> ? Then the Q.resolve(null) can be omitted and the final .then handlers can just return void

File: src/taco-cli/test/kit.ts [hunk 9]
```diff
@@ -198,34 +302,45 @@ describe("Kit", function (): void {
...
-            runKitCommandAndVerifyTacoJsonContents(["select", "--cordova", "5.1.1"], tacoJsonPath, expectedCliTacoJsonKeyValues)
-                .then((telemetryParameters: TacoUtility.ICommandTelemetryProperties) => {
+        it("'taco kit select --cordova {Invalid-CLI-VERSION}' should execute with expected errors", function (done: MochaDone): void {
+            runKitCommandFailureCaseAndVerifyTacoJson<TacoErrorCodes>(["select", "--cordova", "InvalidCordovaCliVersion"], tacoJsonPath, expectedKitTacoJsonKeyValues, TacoErrorCodes.ErrorInvalidVersion)
+                .done(() => done(), done);
+        });
+
+        it("'taco kit select --cordova {CLI-VERSION}' on a project with a platform added, should execute with no errors", function (done: MochaDone): void {
+            /* tslint:disable:no-empty */
+            KitMod.yesOrNoHandler = getMockYesOrNoHandler(done, () => {}, "PromptResponseNo");
```
Comment (MSLaguana) [inline src/taco-cli/test/kit.ts]: We have a TacoUtils.UtilHelper.emptyMethod for intentionally empty functions so you don't need to disable the tslint check around them. Makes it clear that it is intended to be empty

File: src/taco-cli/test/kit.ts [hunk 10]
```diff
@@ -198,34 +302,45 @@ describe("Kit", function (): void {
...
         var tacoJsonPath: string = path.resolve(cliProjectpath, "taco.json");
-        var expectedKitTacoJsonKeyValues: IKeyValuePair<string> = {
-            kit: "5.1.1-Kit", "cordova-cli": "5.1.1"
-        };
 
-        this.timeout(60000);
+        this.timeout(180000);
 
         before(function (done: MochaDone): void {
+            rimraf.sync(cliProjectpath);
```
Comment (MSLaguana) [inline src/taco-cli/test/kit.ts]: will cliProjectpath always exist? Or could this throw an exception if the folder is already absent?

File: src/taco-cli/test/kit.ts [hunk 11]
```diff
@@ -233,16 +348,29 @@ describe("Kit", function (): void {
             rimraf(cliProjectpath, function (err: Error): void { done(); }); // ignore errors
         });
 
-        it("'taco kit select --kit {kit-ID}' should execute with no errors", function (done: MochaDone): void {
-            runKitCommandAndVerifyTacoJsonContents(["select", "--kit", "5.1.1-Kit"], tacoJsonPath, expectedKitTacoJsonKeyValues)
-                .then((telemetryParameters: TacoUtility.ICommandTelemetryProperties) => {
-                    var expected: TacoUtility.ICommandTelemetryProperties = {
+        it("'taco kit select --kit {Invalid-kit-ID}' should execute with expected error", function (done: MochaDone): void {
+            runKitCommandFailureCaseAndVerifyTacoJson<TacoKitsErrorCodes>(["select", "--kit", "InvalidKit"], tacoJsonPath, expectedCliTacoJsonKeyValues2, TacoKitsErrorCodes.TacoKitsExceptionInvalidKit)
```
Comment (MSLaguana) [inline src/taco-cli/test/kit.ts]: A better approach to tests which succeed or fail as a promise chain is to just return the promise, and not accept a done function. Just found this out recently, but it cleans things up a bunch.

This would then be
`it("...", function (): Q.promise<any> { return runKitCommandCaseAndVerifyTacoJson(...);})`

File: src/taco-cli/test/kit.ts [hunk 12]
```diff
@@ -89,19 +110,92 @@ describe("Kit", function (): void {
...
+            return sleep(100);
+        }).then(function (): Q.Promise<any> {
+            var medicPluginTestsPath = path.resolve(".", "plugins", "cordova-plugin-camera", "tests");
+            var pluginCommandArgs: string[] = ["add", medicPluginTestsPath];
+            return pluginRun(pluginCommandArgs);
+        });
+    }
+
+    function getMockYesOrNoHandler(errorHandler: (err: Error) => void, onClose: () => void, desiredResponse: string): {
+        question: (question: string, callback: (answer: string) => void) => void;
```
Comment (MSLaguana) [inline src/taco-cli/test/kit.ts]: This could return an IMockReadLine if you export that from kit.ts

File: src/taco-tests-utils/projectHelper.ts [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+                }
+
+                var cmdPath: string = path.join(projectPath, "platforms", platform, "cordova", cmdName);
+                var versionProc: child_process.ChildProcess = child_process.spawn(cmdPath);
+                versionProc.stdout.on("data", function(data: any): void {
+                    var version: string = data.toString();
+                    if (!path) {
+                        deferred.reject("Verson string not found");
+                    } else {
+                        version.trim().should.be.equal(platformsExpected[platform]);
```
Comment (MSLaguana) [inline src/taco-tests-utils/projectHelper.ts]: This check should probably not be an assertion, or it should be in a try catch, since right now I think this will be an uncaught exception

File: src/taco-tests-utils/projectHelper.ts [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+                });
+                versionProc.on("error", function(err: any): void {
+                    deferred.reject(err);
+                });
+                return deferred.promise;
+            }));
+        }
+
+        public static checkPluginVersions(pluginsExpected: IKeyValuePair<string>, projectPath: string): void {
+            var deferred: Q.Deferred<any> = Q.defer<any>();
```
Comment (MSLaguana) [inline src/taco-tests-utils/projectHelper.ts]: This deferred isn't used

File: src/taco-cli/test/kit.ts (hunk 12 above)
Comment (meenakb) [inline src/taco-cli/test/kit.ts]: Addressed the comments and added comments where appropriate.

File: src/taco-tests-utils/projectHelper.ts [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+module TacoTestsUtils {
+    interface IKeyValuePair<T> {
+        [key: string]: T;
+    }
+
+    export class ProjectHelper {
+        public static checkPlatformVersions(platformsExpected: IKeyValuePair<string>, projectPath: string): Q.Promise<any> {
+            var platformsInstalled: string[] = ProjectHelper.getInstalledPlatforms(platformsExpected, projectPath);
+            var onWindows: boolean = process.platform === "win32";
+            var deferred: Q.Deferred<any> = Q.defer<any>();
```
Comment (MSLaguana) [inline src/taco-tests-utils/projectHelper.ts]: I think this deferred should be inside the platformsInstalled.map function, otherwise as soon as the first platform finishes everything will be considered done (since there is only one deferred, returned multiple times)

**Comment by MSLaguana** [comment]
Looks good other than that.
